modules/dnn/src/layers/gather_layer.cpp

```diff
@@ -45,34 +45,70 @@ class GatherLayerImpl CV_FINAL : public GatherLayer
         CV_TRACE_FUNCTION();
         CV_TRACE_ARG_VALUE(name, "name", name.c_str());
 
+        // FP16 fallback is not needed as we handle FP16 below
+
         std::vector<Mat> inputs, outputs;
         inputs_arr.getMatVector(inputs);
         outputs_arr.getMatVector(outputs);
 
+        CV_CheckEQ(inputs.size(), (size_t)2, "");
+        CV_CheckEQ(outputs.size(), (size_t)1, "");
+
         const Mat& inp = inputs[0];
-        const Mat& indices = inputs[1];
+
+        int indicesType = inputs[1].type();
+        CV_CheckType(indicesType, indicesType == CV_32FC1 || indicesType == CV_16SC1, "");
+        Mat indices32S;
+        if (indicesType == CV_16S/*FP16*/)
+        {
+            Mat indicesF32;
+            convertFp16(inputs[1], indicesF32);
+            indicesF32.convertTo(indices32S, CV_32S);
+        }
+        else
+        {
+            inputs[1].convertTo(indices32S, CV_32S);
+        }
+        const size_t indices_total = indices32S.total();
+        indices32S = indices32S.reshape(1, indices_total);
+
         Mat& out = outputs[0];
 
+        CV_CheckTypeEQ(inp.type(), out.type(), "");
+        CV_CheckTypeEQ(indices32S.type(), CV_32SC1, "");
+
         const int axis = normalize_axis(m_axis, shape(inp));
 
+        // FIXIT: why should we work with non-normalized input? it should be handled in importer or layers's output generator
```

Comment:
It was originally normalized on the go:

```cpp
for (size_t i = 0; i < outer_dims; ++i)
{
  // ...
  for (size_t j = 0 ; j < indices.total(); ++j)
  {
    const size_t index = (static_cast<int>(idx[j]) + inp.size[axis]) % inp.size[axis];
    // ..
  }
}
```

I think in onnx importer we should load what it is without any extra operation on constant input & attributes. Operations like normalization should be done in the layer constructor or somewhere. We need the original information when it comes to build operators for other backend, like CANN and TIM-VX, who have operator primitives aligned with ONNX, TF ...

Reply:
"On the go" normalization is not efficient as performs it multiple times (especially through division).
There is change to perform normalization once (per forward() call) before the main loop.

Reply:
It was done because the elements of index tensor can be negative per spec and non-constant at the same time.